### DISCUSSION TITLE
cargo vendor not work

### DISCUSSION THREAD

**Discussion by Git-Blue** [CeresDB/ceresdb#192] [Q&A] [answered]
I use `cargo vendor` command to solve dependencies and build offline, then it shows below error message
my host is Mac, ceresdb version 0.2 

```
error: failed to sync

Caused by:
  found duplicate version of package `ceresdbproto v0.1.0` vendored from two sources:

  	source 1: https://github.com/CeresDB/ceresdbproto.git?rev=2c10152d021cd5a26b9c870cdede6a0317adca3d#2c10152d
  	source 2: https://github.com/CeresDB/ceresdbproto.git#a7db8811
```

someone can help?

**Comment by jiacai2050** [comment]
Could you try latest main branch?

**Reply by Git-Blue** [reply]
en.., I tried `cargo vendor` on master branch, It failed
```
error: failed to sync

Caused by:
  found duplicate version of package `ceresdbproto v0.1.0` vendored from two sources:

  	source 1: https://github.com/CeresDB/ceresdbproto.git?rev=2c10152d021cd5a26b9c870cdede6a0317adca3d#2c10152d
  	source 2: https://github.com/CeresDB/ceresdbproto.git?rev=ec5404e0f60fc748d3e8b68a592e2926ac1e84de#ec5404e0
```

**Reply by jiacai2050** [reply]
I have found the cause, SQL module depend directly on ceresdbproto now, change to [ceresdbproto_deps](https://github.com/CeresDB/ceresdb/blob/main/ceresdbproto_deps/Cargo.toml) should fix this, would you like to give it a try?

- https://github.com/CeresDB/ceresdb/blob/653d279a890f0284b3e97e3c9bb35f4f2509ba92/sql/Cargo.toml#L26

**Reply by Git-Blue** [reply]
I find two crates - sql and server both own this  `ceresdbproto` in their Cargo.toml file, and direct replace 

>ceresdbproto = { git = "https://github.com/CeresDB/ceresdbproto.git", rev = "2c10152d021cd5a26b9c870cdede6a0317adca3d" } 

with 

> ceresdbproto_deps = { path = "../ceresdbproto_deps" }

do not work , because the `root crate` is `ceresdbproto_deps::`, so should modify `src` code for both crates. I do not prefer to do that change, so I just replace the `ceresdbproto` revision num to newer `ec5404e0f60fc748d3e8b68a592e2926ac1e84de` to keep consistent and It works. I am not sure whether those two commit diff too much, or maybe you can offer better solution.
beside, I find two similar depends `https://github.com/breeswish/slog-global.git` and `https://github.com/breezewish/slog-global.git` in `cargo.lock` file, both direct to `https://github.com/tikv/slog-global`, It is not a mistake, but confusing.

**Reply by ShiKaiWi** [reply]
> I am not sure whether those two commit diff too much, or maybe you can offer better solution.
beside, I find two similar depends https://github.com/breeswish/slog-global.git and https://github.com/breezewish/slog-global.git in cargo.lock file, both direct to https://github.com/tikv/slog-global, It is not a mistake, but confusing.

Thanks for reporting this **implicit** weird issue and I will address it.

**Reply by jiacai2050** [reply]
https://github.com/CeresDB/ceresdb/pull/195 fix slog issue.

**Comment by chunshao90** [comment, marked as answer]
Refer to #190 , this pr will fix.
